pkg/cloudprovider/provider/gce/service.go

```diff
@@ -72,16 +80,66 @@ func (svc *service) networkInterfaces(cfg *config) ([]*compute.NetworkInterface,
 		Subnetwork: cfg.subnetwork,
 	}
 
+	klog.Infof("using network:%s subnetwork: %s", cfg.network, cfg.subnetwork)
+
 	if cfg.assignPublicIPAddress {
-		ifc.AccessConfigs = []*compute.AccessConfig{{
-			Name: "External NAT",
-			Type: "ONE_TO_ONE_NAT",
-		}}
+		ifc.AccessConfigs = []*compute.AccessConfig{
+			{
+				Name: "External NAT",
+				Type: "ONE_TO_ONE_NAT",
+			},
+		}
+	}
+
+	// Setup IPv6
+	// GCP allocates public IPv6 addr so we only try to setup IPv6
+	// if assigning public IP addresses is enabled.
+	if cfg.assignPublicIPAddress {
+
+		// GCP doesn't support IPv6 only stack
+		if util.ContainsCIDR(networkConfig.PodCIDRs, util.IPv4) &&
+			util.ContainsCIDR(networkConfig.PodCIDRs, util.IPv6) {
+
+			if isIPv6Suported(cfg.zone) {
+				ifc.StackType = "IPV4_IPV6"
+
+				ifc.Ipv6AccessConfigs = []*compute.AccessConfig{
+					{
+						Name:        "external-ipv6",
+						NetworkTier: "PREMIUM", // TODO: check if ipv6 is supported in other tiers
+						Type:        "DIRECT_IPV6",
+					},
+				}
+			} else {
+				return nil, fmt.Errorf("dualstack is not supported in %q zone", cfg.zone)
+			}
+
+		} else {
+			klog.Infof("pod cidr %q doesn't specify dual stack cluster", networkConfig.PodCIDRs)
+		}
 	}
 
 	return []*compute.NetworkInterface{ifc}, nil
 }
 
+func isIPv6Suported(zone string) bool {
```

Comment:
This check should be added in the webhook in the `Vaildate` method. So that you validate the support for IPv6 before you create the Machine

Reply:
Also there is a typo s/isIPv6Suported/isIPv6Supported